File: internal/webui/devui/app/api/logger.ts
// Copyright 2022 Namespace Labs Inc; All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

export const InfoLevel = 1;
export const DebugLevel = 2;

export class Logger {
	private readonly what: string;
	private level: number;

	constructor(what: string, level: number = InfoLevel) {
		this.what = what;
		this.level = level;
	}

	info(...args: any[]) {
		if (InfoLevel <= this.level) {
			console.log(`info[${this.what}]`, ...args);
		}
	}

	debug(...args: any[]) {
		if (DebugLevel <= this.level) {
			console.log(`debug[${this.what}]`, ...args);
		}
	}

	error(...args: any[]) {
		console.error(`error[${this.what}]`, ...args);
	}
}

File: internal/console/logger.go
// Copyright 2022 Namespace Labs Inc; All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

package console

import (
	"io"
	"log"

	"gopkg.in/natefinch/lumberjack.v2"
)

type rotatedDebugLogger struct {
	rotatedFile *lumberjack.Logger
	logger      *log.Logger
}

func newRotatedDebugLogger(filePath string) io.WriteCloser {
	rotatedFile := &lumberjack.Logger{
		Filename:   filePath,
		MaxSize:    10, // megabytes
		MaxBackups: 3,
		MaxAge:     10,   // days
		Compress:   true, // disabled by default
	}

	return &rotatedDebugLogger{
		rotatedFile: rotatedFile,
		logger:      log.New(rotatedFile, "", log.LstdFlags|log.Lmicroseconds),
	}
}

func (r *rotatedDebugLogger) Write(buf []byte) (int, error) {
	r.logger.Print(string(buf))
	return len(buf), nil
}

func (r *rotatedDebugLogger) Close() error {
	return r.rotatedFile.Close()
}
